module/zfs/spa_errlog.c

```diff
@@ -128,6 +243,296 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+			}
+			check_snapshot = B_FALSE;
+		}
+	}
+
+	/* How many snapshots reference this block. */
+	uint64_t snap_count;
+	if (zap_count(spa->spa_meta_objset,
+	    dsl_dataset_phys(ds)->ds_snapnames_zapobj, &snap_count) != 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (SET_ERROR(EFAULT));
+	}
+
+	if (snap_count == 0) {
+		/* File system has no snapshot. */
+		dsl_dataset_rele(ds, FTAG);
+		return (0);
+	}
+
+	uint64_t *snap_obj_array = kmem_alloc(snap_count * sizeof (uint64_t),
+	    KM_SLEEP);
+
+	int aff_snap_count = 0;
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	uint64_t zap_clone = dsl_dir_phys(ds->ds_dir)->dd_clones;
+
+	/* Check only snapshots created from this file system. */
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg &&
+	    snap_obj_txg <= txg_to_consider) {
+
+		dsl_dataset_rele(ds, FTAG);
+		if (dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds) != 0)
+			return (SET_ERROR(EFAULT));
+
+		if (dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj != fs)
+			break;
+
+		boolean_t affected = B_TRUE;
+		if (check_snapshot) {
+			affected = B_FALSE;
+			uint64_t blk_txg;
+			if (find_block_txg(ds, zep, &blk_txg) == 0 &&
+			    zep->zb_birth == blk_txg) {
+				affected = B_TRUE;
+			}
+		}
+
+		if (affected) {
+			snap_obj_array[aff_snap_count] = snap_obj;
+			aff_snap_count++;
+
+			if (!only_count) {
+				zbookmark_phys_t zb;
+				zep_to_zb(snap_obj, zep, &zb);
+				if (copyout(&zb, (char *)addr + (*count - 1)
+				    * sizeof (zbookmark_phys_t),
+				    sizeof (zbookmark_phys_t)) != 0) {
+					dsl_dataset_rele(ds, FTAG);
+					goto out;
+				}
+				(*count)--;
+			} else {
+				(*count)++;
+			}
+		}
+		snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+		snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	}
+	dsl_dataset_rele(ds, FTAG);
+
+	/* Check clones referencing the error block. */
+	if (zap_clone != 0 && aff_snap_count > 0) {
+		zap_cursor_t zc;
+		zap_attribute_t za;
+
+		for (zap_cursor_init(&zc, spa->spa_meta_objset, zap_clone);
+		    zap_cursor_retrieve(&zc, &za) == 0;
+		    zap_cursor_advance(&zc)) {
+
+			dsl_pool_t *dp = spa->spa_dsl_pool;
+			dsl_dataset_t *clone;
+			int err = dsl_dataset_hold_obj(dp, za.za_first_integer,
+			    FTAG, &clone);
+
+			if (err != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+
+			boolean_t found = B_FALSE;
+			for (int i = 0; i < aff_snap_count; i++) {
+				if (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+				    == snap_obj_array[i]) {
+					found = B_TRUE;
+				}
+			}
+			dsl_dataset_rele(clone, FTAG);
+
+			if (!found)
+				continue;
+
+			err = check_filesystem(spa, za.za_first_integer, zep,
+			    count, addr, only_count);
+
+			if (err != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+		}
+		zap_cursor_fini(&zc);
+	}
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (0);
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (SET_ERROR(EFAULT));
```

Comment:
How about a single exit path which returns `err`, which can be set before the `goto`.

```
out:
	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
	return (err);
``